Fix leaking LeaseRejectedError

mrtracy opened the pull request:
In a very particular race condition, a LeaseRejectedError could leak back to
clients. This race occurs when a replica requests the leader lease, but it is no
longer part of the range's raft group (having not yet been GCed). Due to a logic
quirk, this was not being properly converted to a "NotLeaderError" before being
passed back to the sender level.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/2996)

<!-- Reviewable:end -->

tschottdorf commented on storage/replica.go:
```diff
@@ -381,14 +381,19 @@ func (r *Replica) redirectOnOrAcquireLeaderLease(trace *tracer.Trace, timestamp
+			log.Warningf("Lease for range %s rejected at timestamp %d: %s", r, timestamp,
+				err)
 		}
+		return r.newNotLeaderError(r.getLease(), r.store.StoreID())
```
can you double-check how that lease is interpreted by the caller? In the "new" case here, our lease is possibly outdated, which I'm not sure the caller knows. Might be better to return a lease-less `NewLeaderError` then.

mrtracy replied:
I restored the original logic which did just this; however, there is now an `else` case which returns a leaderless NotLeaderError in other cases.

tschottdorf commented:
Can you add a simple test case that exercises this new change?

tamird commented on storage/replica.go:
```diff
@@ -381,14 +381,19 @@ func (r *Replica) redirectOnOrAcquireLeaderLease(trace *tracer.Trace, timestamp
-		if lease := r.getLease(); lease.Covers(timestamp) {
-			return r.newNotLeaderError(lease, r.store.StoreID())
+		if log.V(1) {
+			log.Warningf("Lease for range %s rejected at timestamp %d: %s", r, timestamp,
```
can't print a timestamp as `%d`

tamird replied:
(how did this pass `vet`?)

bdarnell replied:
I don't think vet knows about `Warningf` by default.

tamird replied:
Ah, that must be it.

mrtracy replied:
Changed to %v

Commit:
Fixing LeaseRejectedError leak

In a very particular race condition, a LeaseRejectedError could leak back to
clients. This race occurs when a replica requests the leader lease, but it is no
longer part of the range's raft group (having not yet been GCed). Due to a logic
quirk, this was not being properly converted to a "NotLeaderError" before being
passed back to the sender level.

mrtracy commented:
Added a simple test to exercise this.

bdarnell commented:
LGTM

tschottdorf commented on storage/replica.go:
```diff
@@ -381,14 +381,23 @@ func (r *Replica) redirectOnOrAcquireLeaderLease(trace *tracer.Trace, timestamp
 	if _, ok := err.(*roachpb.LeaseRejectedError); ok {
-		if lease := r.getLease(); lease.Covers(timestamp) {
-			return r.newNotLeaderError(lease, r.store.StoreID())
+		if log.V(1) {
```
should this be the `else if` for line 397 below? After all, it's expected that this happen when multiple replicas race for the lease.

mrtracy replied:
Yeah, good point.  Moved.

tamird commented on storage/replica_test.go:
```diff
@@ -487,6 +487,20 @@ func TestRangeLeaderLease(t *testing.T) {
 	}
+
+	// Verify that command returns NotLeaderError when lease is rejected.
+	tc.rng.proposeRaftCommandFn = func(id cmdIDKey, cmd roachpb.RaftCommand) <-chan error {
```
is it safe to do this without locking in the middle of a test?

tschottdorf replied:
if this range is part of a store (and it seems it is), then this could race since the store will periodically make use of its first range. You'd have to set it earlier (check how the other tests that use this mock are doing it).

mrtracy replied:
The other tests create a new replica, which is apparently easily done.  The new test now does the same, there shouldn't be a risk of race anymore.

tschottdorf commented:
LGTM

tamird commented on storage/replica_test.go:
```diff
@@ -487,6 +487,21 @@ func TestRangeLeaderLease(t *testing.T) {
 	}
+
+	// Verify that command returns NotLeaderError when lease is rejected.
+	rng, err := NewReplica(testRangeDescriptor(), tc.store)
```
probably should do something with this error.

mrtracy replied:
Done

tamird commented on storage/replica.go:
```diff
@@ -381,14 +381,25 @@ func (r *Replica) redirectOnOrAcquireLeaderLease(trace *tracer.Trace, timestamp
+			// The lease was rejected even though it was not obtained by another
+			// replica.
+			if log.V(1) {
+				log.Warningf("Lease for range %s rejected at timestamp %v: %s", r, timestamp,
```
since you'll have to amend this anyway, make this line shorter

mrtracy replied:
Done

tamird commented on storage/replica_test.go:
```diff
@@ -487,6 +487,24 @@ func TestRangeLeaderLease(t *testing.T) {
+	}
+
+	err = rng.redirectOnOrAcquireLeaderLease(nil, tc.clock.Now())
+	if lErr, ok := err.(*roachpb.NotLeaderError); !ok || lErr == nil {
```
the `lErr == nil` check is not useful; it's not possible for `ok` to be true and `lErr` to be nil at the same time. I'll send a PR with other fixes and grab this.

tschottdorf replied:
isn't that the "it's possible, but we've decided that it's not something we'll ever have" case?

```go
var err error
err = (*roachpb.NotLeaderError)(nil)
```

tamird replied:
Yeah, that's true. Still, it seems to be in alignment with our practices to avoid this check. For the curious: http://play.golang.org/p/tLd4XGxKjQ

tschottdorf replied:
agreed. We had this discussion before.
